### PR TITLE
Support sets as arguments to functions

### LINKED ISSUE #1663: [SETS] Support sets as arguments in C
### Describe the feature
Support passing sets as arguments to C code

### Test Code
Provide code which does not currently work but which should do when this issue is fixed:
```python
def f(a : set[int]):
    pass
```

### Proposed Solution
The cwrapper should be improved to ensure that sets can be read from Python and passed to C code.

### Prerequisites
#1659

### PR DESCRIPTION
**Pull request by EmilyBourne**
Improve the C-Python wrapper to allow passing sets as arguments. Fixes #1663

**Commit Summary**
- Add documentation for set annotation
- Add description for C-Python API set manipulation functions
- Add support for constant set arguments to C functions
- Remove include guards from `Set_extensions.h` to allow extensions to multiple set types
- Add a test for a function with a constant set argument

### CODE REVIEW

**Comment by EmilyBourne** [comment]
Blocked by #2030

**Review by jalalium — CHANGES_REQUESTED** [review]

File: pyccel/codegen/wrapper/c_to_python_wrapper.py [hunk 1]
```diff
@@ -2491,4 +2520,89 @@ def _extract_HomogeneousTupleType_FunctionDefArgument(self, orig_var, collect_ar
...
+        return {'body': body, 'args': arg_vars}
+
+    def _extract_HomogeneousSetType_FunctionDefArgument(self, orig_var, collect_arg, bound_argument,
+            is_bind_c_argument, *, arg_var = None):
+        """
+        Extract the C-compatible homogeneous tuple FunctionDefArgument from the PythonObject.
+
+        Extract the C-compatible homogeneous tuple FunctionDefArgument from the PythonObject.
+        The C-compatible argument is extracted from collect_arg which holds a Python
+        oject into arg_var.
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/codegen/wrapper/c_to_python_wrapper.py]: Fixed [2ae9b55](https://github.com/pyccel/pyccel/pull/2031/commits/2ae9b5501a7205557100861efdbf48d804f8e7df)

**Review by jalalium — APPROVED** [review]
Well done!

**Comment by EmilyBourne** [comment]
Blocked by #2070

File: pyccel/codegen/wrapper/c_to_python_wrapper.py [hunk 2]
```diff
@@ -2352,49 +2426,100 @@ def _extract_HomogeneousTupleType_FunctionDefArgument(self, orig_var, collect_ar
...
-            translated code is in C.
+        orig_var : Variable | IndexedElement
+            An object representing the variable or an element of the variable from the
+            FunctionDefArgument being wrapped.
+
+        collect_arg : Variable
+            A variable with type PythonObject* holding the Python argument from which the
+            C-compatible argument should be collected.
+
+        bound_argument : bool
```
Comment (jalalium) [inline pyccel/codegen/wrapper/c_to_python_wrapper.py:2451]: What's the point of this argument if it is always false?

Comment (EmilyBourne) [reply pyccel/codegen/wrapper/c_to_python_wrapper.py:2451]: All the `_extract_X_FunctionDefArgument` functions need to have the same arguments. In `_extract_CustomDataType_FunctionDefArgument` I need to know if the argument is a bound argument or not so this is added as an argument even if it is unhelpful for other functions. This is out of scope for this PR but if you have a suggestion for how to get this information in `_extract_CustomDataType_FunctionDefArgument` without passing it as an argument then we should create an issue and fix this in a dedicated PR.

Comment (jalalium) [reply pyccel/codegen/wrapper/c_to_python_wrapper.py:2451]: Thank you for the clarification! No, I just didn't get it

**Pull request merged**